### PR TITLE
Fix local engine use in macOS plugins

### PR DESCRIPTION
**Pull request by stuartmorgan-g**
Currently podhelper.rb will always point plugin builds at the cached engine artifacts, even when using `--local-engine`. In most cases this is fine, since when the final build actually runs it will be using the engine bundled into the app build, which will be the correct local engine build. When trying to test a local engine build with API additions against a local plugin modified to use those additions to ensure that they are working as expected, however, compilation will fail, because the new APIs won't be present in the plugin build.

This fixes that for macOS, and adds a TODO for iOS (which is more complicated to fix due to the host vs target build distinction).

macOS portion of https://github.com/flutter/flutter/issues/132228

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: dev/devicelab/lib/tasks/plugin_tests.dart [hunk 1]
```diff
@@ -84,6 +84,12 @@ class PluginTest {
           section('Test app');
           await app.runFlutterTest();
         }
+        // Validate local engine handling. Currently only implemented for macOS.
```
Comment (stuartmorgan) [inline dev/devicelab/lib/tasks/plugin_tests.dart:87]: Alternately I could make this a new top-level `PluginTest` parameter and run it separately in the various plugin test tasks, but it's so fast to run this seemed easier.

File: packages/flutter_tools/bin/podhelper.rb [hunk 2]
```diff
@@ -75,6 +75,7 @@ def flutter_additional_ios_build_settings(target)
     build_configuration.build_settings['ENABLE_BITCODE'] = 'NO'
 
     # Profile can't be derived from the CocoaPods build configuration. Use release framework (for linking only).
+    # TODO(stuartmorgan): Handle local engines here; see https://github.com/flutter/flutter/issues/132228
```
Comment (stuartmorgan) [inline packages/flutter_tools/bin/podhelper.rb:78]: Originally I was going to fix both, but iOS is messier and I don't know when I'll get back to it, so I decided to at least land the macOS part and document the pitfall before I lose track of the branch where I had fixed macOS.

File: packages/flutter_tools/bin/podhelper.rb [hunk 3]
```diff
@@ -319,3 +324,36 @@ def flutter_relative_path_from_podfile(path)
   relative = pathname.relative_path_from project_directory_pathname
   relative.to_s
 end
+
+def flutter_parse_xcconfig_file(file)
```
Comment (stuartmorgan) [inline packages/flutter_tools/bin/podhelper.rb:328]: Sorry to have to bring this code back :(

**Comment by stuartmorgan** [comment]
Weird, I had that test working locally. Maybe it only works in some variants of the `PluginTest` invocation; I'll have to run all of them again.

**Review by jmagman — APPROVED** [review]
The podhelper change LGTM, I'll re-review when you get the test working.

**Comment by stuartmorgan** [comment]
Turns out doing a `build --config-only` when you already have a complete build doesn't regenerate the pod bits. It worked for me before because I had temporarily commented out the earlier test steps to make the write/test/debug iteration cycle faster.

I've added a `clean` before generating the local-engine config.

File: packages/flutter_tools/bin/podhelper.rb [hunk 4]
```diff
@@ -319,3 +324,36 @@ def flutter_relative_path_from_podfile(path)
...
+  entries = Hash.new
+  skip_line_start_symbols = ["#", "/"]
+  File.foreach(file_abs_path) { |line|
+      next if skip_line_start_symbols.any? { |symbol| line =~ /^\s*#{symbol}/ }
+      key_value_pair = line.split(pattern='=')
+      if key_value_pair.length == 2
+        entries[key_value_pair[0].strip()] = key_value_pair[1].strip();
+      else
+        puts "Invalid key/value pair: #{line}"
+      end
```
Comment (jmagman) [inline packages/flutter_tools/bin/podhelper.rb:342]: I think the whitespace is off here.
```suggestion
    next if skip_line_start_symbols.any? { |symbol| line =~ /^\s*#{symbol}/ }
    key_value_pair = line.split(pattern = '=')
    if key_value_pair.length == 2
      entries[key_value_pair[0].strip()] = key_value_pair[1].strip();
    else
      puts "Invalid key/value pair: #{line}"
    end
```

**Review by jmagman — APPROVED** [review]
LGTM

**Pull request merged**